d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/FailedoutClusterManager.java

```diff
@@ -0,0 +1,169 @@
+   */
+  public void updateFailoutConfig(@Nullable FailoutConfig failoutConfig)
+  {
+    if (failoutConfig == null)
+    {
+      removePeerClusterWatches();
+    }
+    else
+    {
+      processNewConfig(failoutConfig);
+    }
+
+    _failoutConfig = failoutConfig;
+  }
+
+  private void processNewConfig(@Nonnull FailoutConfig failoutConfig)
+  {
+    if (!failoutConfig.isFailedOut())
+    {
+      _log.debug("Failout completed for cluster: {}. Removing all peer cluster watches.", _clusterName);
+      removePeerClusterWatches();
+    }
+    else
+    {
+      Set<String> peerClusters = failoutConfig.getPeerClusters();
+      addPeerClusterWatches(peerClusters);
+    }
+  }
+
+  /**
+   * Calle this method when a cluster is failed out and/or its new peer clusters are identified.
+   * @param newPeerClusters Name of the peer clusters of the failed out clusters
+   */
+  void addPeerClusterWatches(@Nonnull Set<String> newPeerClusters)
+  {
+    final Set<String> existingPeerClusters = _clusterListeners.keySet();
+
+    if (newPeerClusters.isEmpty())
+    {
+      removePeerClusterWatches();
+      return;
+    }
+
+    final Set<String> peerClustersToAdd = new HashSet<>(newPeerClusters);
+    peerClustersToAdd.removeAll(existingPeerClusters);
+
+    if (!peerClustersToAdd.isEmpty())
+    {
+      addClusterWatches(peerClustersToAdd);
+    }
+
+    final Set<String> peerClustersToRemove = new HashSet<>(existingPeerClusters);
+    peerClustersToRemove.removeAll(newPeerClusters);
+
+    if (!peerClustersToRemove.isEmpty())
+    {
+      removeClusterWatches(peerClustersToRemove);
+    }
+  }
+
+  /**
+   * Call this method when a cluster failed out is over and we do not need to monitor its peer clusters.
+   */
+  void removePeerClusterWatches()
+  {
+    removeClusterWatches(_clusterListeners.keySet());
+  }
+
+  private void addClusterWatches(@Nonnull Set<String> clustersToWatch)
+  {
+    if (_log.isDebugEnabled())
+    {
+      _log.debug("Watching peer clusters: " + String.join(",", clustersToWatch));
+    }
+    for (final String cluster : clustersToWatch)
+    {
+      _clusterListeners.computeIfAbsent(cluster, clusterName -> {
+        // TODO(RESILIEN-50): Establish connections to peer clusters when listener#done() is invoked.
+        LoadBalancerStateListenerCallback listener = new LoadBalancerState.NullStateListenerCallback();
+        _loadBalancerState.listenToCluster(cluster, listener);
+        return listener;
+      });
+    }
+  }
+
+  private void removeClusterWatches(@Nonnull Set<String> clustersToRemove)
+  {
+    if (_log.isDebugEnabled())
+    {
+      _log.debug("Removing peer clusters: " + String.join(",", clustersToRemove));
+    }
+    for (String cluster : clustersToRemove)
+    {
+      final LoadBalancerStateListenerCallback listener = _clusterListeners.remove(cluster);
+      if (listener != null)
+      {
+        // TODO(RESILIEN-51): Unregister watches when failout is over
```

Comment:
What's the plan for completing the TODOs here? with follow-up PRs?

Reply:
Yes, those will be in follow-up PRs.